### PR TITLE
Improve region endpoint resolver

### LINKED ISSUE #410: kiam fails to resolve regional sts endpoint in us-iso-east-1 and us-isob-east-1
Hello!

We're deploying kiam in two of AWS's air gapped regions and we're finding that the sts endpoint fails to resolve.  

In us-iso-east-1, we're falling to this statement:  https://github.com/uswitch/kiam/blob/master/pkg/aws/sts/gateway.go#L73
And getting the error
```
Regional STS endpoint does not exist: sts.us-iso-east-1.amazonaws.com
```

After peeking around the code, I think we need special handling for these regions the same way china has special handling like here https://github.com/uswitch/kiam/blob/master/pkg/aws/sts/gateway.go

``` 
	if strings.HasPrefix(region, "cn-") {
		host = fmt.Sprintf("sts.%s.amazonaws.com.cn", region)
	} else if strings.HasPrefix(region, "us-isob") {
                host = fmt.Sprintf("sts.%s.sc2s.sgov.gov", region)
	} else if strings.HasPrefix(region, "us-iso") {
                host = fmt.Sprintf("sts.%s.c2s.ic.gov", region)
        } else {
		host = fmt.Sprintf("sts.%s.amazonaws.com", region)
	}
```

The Go AWS SDK handles these regions itself [(endpoints.json)](https://github.com/aws/aws-sdk-go-v2/blob/master/models/endpoints/endpoints.json#L8193-L8203) though so I wonder if there isn't some other issue?  I'm not familiar enough with the code to know for sure so sorry if this is some larger issue.  Thank you for any help you can provide!

### PR DESCRIPTION
**Pull request by pingles**
Pull out into separate type and add some more tests about handling of gov, fips and chinese regions.

Additionally removes comparison against SDK regions and instead relies on DNS resolution to verify.

This touches on, but doesn't necessarily fix #410 #368 #344 yet.

### CODE REVIEW

**Comment by pingles** [comment]
https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_enable-regions.html#id_credentials_temp_enable-regions_writing_code has more context on setting regions with STS

File: pkg/aws/sts/aws_endpoint_resolver.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+	if _, err := net.LookupHost(hostname); err != nil {
+		return "", fmt.Errorf("Regional STS endpoint does not exist: %s", hostname)
+	}
+
+	return hostname, nil
+}
+
+func newRegionalEndpointResolver(region string) (endpoints.Resolver, error) {
+	if region == "" || strings.Contains(region,"fips") {
```
Comment (Joseph-Irving) [inline pkg/aws/sts/aws_endpoint_resolver.go:51]: worth adding a comment about why this is the case

File: pkg/aws/sts/aws_endpoint_resolver.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+func regionalHostname(region string) (string, error) {
+	hostname := fmt.Sprintf("sts.%s.amazonaws.com", region)
+
+	if strings.HasPrefix(region, "cn-") {
+		hostname = fmt.Sprintf("%s.cn", hostname)
+	}
+
+	if strings.HasPrefix(region, "us-iso") {
+		hostname = fmt.Sprintf("sts.%s.c2s.ic.gov", region)
+	}
```
Comment (Joseph-Irving) [inline pkg/aws/sts/aws_endpoint_resolver.go:41]: probably worth a comment explaining why the difference here

File: pkg/aws/sts/aws_endpoint_resolver.go (hunk 1 above)
Comment (pingles) [reply pkg/aws/sts/aws_endpoint_resolver.go:58]: Added a note in the latest

File: pkg/aws/sts/aws_endpoint_resolver.go (hunk 2 above)
Comment (pingles) [reply pkg/aws/sts/aws_endpoint_resolver.go]: Thanks, added a note in the latest commit

**Comment by pingles** [comment]
Latest, this definitely fixes #410. 

It should also fix #368 which would have accessed the global endpoint while retrieving server credentials (and then using regional endpoints for pod credentials). The tests show it using a regional resolver and how it resolves the endpoints.

**Review by Joseph-Irving — APPROVED** [review]
lgtm

**Pull request merged**